crates/wast/src/resolve/component_names.rs

```diff
@@ -0,0 +1,379 @@
+                        resolve_type_use(&mut r.ok, resolver)?;
+                        resolve_type_use(&mut r.err, resolver)?;
+                    }
+                }
+                Ok(())
+            }
+        },
+
+        ComponentField::Func(_f) => Ok(()),
+
+        ComponentField::Instance(_i) => Ok(()),
+        ComponentField::Module(_m) => Ok(()),
+        ComponentField::Component(_c) => Ok(()),
+        ComponentField::Alias(_a) => todo!("register for alias"),
+
+        ComponentField::Start(_i) => Ok(()),
+
+        ComponentField::Export(_e) => Ok(()),
+
+        ComponentField::Custom(_) => Ok(()),
+    }
+}
+
+fn resolve_type_use<'a, 'b, T>(
+    ty: &'b mut ComponentTypeUse<'a, T>,
+    resolver: &'b mut ComponentResolver<'a>,
+) -> Result<(), Error> {
+    match ty {
+        ComponentTypeUse::Ref(r) => resolve_item_ref(r, resolver),
+        ComponentTypeUse::Inline(_i) => {
+            // Nothing to do.
+            Ok(())
+        }
+    }
+}
+
+fn resolve_item_ref<'a, 'b, K>(
+    item: &'b mut ItemRef<'a, K>,
+    resolver: &'b mut ComponentResolver<'a>,
+) -> Result<(), Error>
+where
+    K: Into<Ns> + Copy,
+{
+    #[cfg(wast_check_exhaustive)]
+    {
+        if !item.visited {
+            return Err(Error::new(
+                item.idx.span(),
+                format!("BUG: this index wasn't visited"),
+            ));
+        }
+    }
+    resolve_ns(&mut item.idx, item.kind.into(), resolver)?;
+    Ok(())
+}
+
+fn resolve_ns<'a, 'b>(
+    idx: &mut Index<'a>,
+    ns: Ns,
+    resolver: &'b mut ComponentResolver<'a>,
+) -> Result<u32, Error> {
+    match ns {
+        Ns::Func => resolver.funcs.resolve(idx, "func"),
+        Ns::Table => resolver.tables.resolve(idx, "table"),
+        Ns::Global => resolver.globals.resolve(idx, "global"),
+        Ns::Memory => resolver.memories.resolve(idx, "memory"),
+        Ns::Tag => resolver.tags.resolve(idx, "tag"),
+        Ns::Type => resolver.types.resolve(idx, "type"),
+        Ns::Component => resolver.components.resolve(idx, "component"),
+        Ns::Module => resolver.modules.resolve(idx, "module"),
+        Ns::Instance => resolver.instances.resolve(idx, "instance"),
+        Ns::Value => resolver.values.resolve(idx, "instance"),
+    }
+}
+
+#[derive(PartialEq, Eq, Hash, Copy, Clone, Debug)]
+enum Ns {
+    Func,
+    Table,
+    Global,
+    Memory,
+    Tag,
+    Type,
+    Component,
+    Module,
+    Instance,
+    Value,
+}
+
+macro_rules! component_kw_conversions {
+    ($($kw:ident => $kind:ident)*) => ($(
+        impl From<kw::$kw> for Ns {
+            fn from(_: kw::$kw) -> Ns {
+                Ns::$kind
+            }
+        }
+    )*);
+}
+
+component_kw_conversions! {
+    func => Func
+    module => Module
+    component => Component
+    instance => Instance
+    value => Value
+    table => Table
+    memory => Memory
+    global => Global
+    tag => Tag
+    r#type => Type
+}
+
+#[derive(Default)]
+struct Namespace<'a> {
```

Comment:
This looks pretty similar to the `Namespace` helper for modules, so could that be reused?